core/src/transport.rs

```diff
@@ -221,149 +228,276 @@ pub trait Transport {
     }
 }
 
-/// Event produced by [`Transport::Listener`]s.
+/// The ID of a single listener.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub struct ListenerId(u64);
+
+impl ListenerId {
+    /// Creates a new `ListenerId`.
+    pub fn new() -> Self {
+        ListenerId(rand::random())
+    }
+}
+
+impl Default for ListenerId {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Event produced by [`Transport`]s.
 ///
-/// Transports are expected to produce `Upgrade` events only for
+/// Transports are expected to produce `Incoming` events only for
 /// listen addresses which have previously been announced via
 /// a `NewAddress` event and which have not been invalidated by
 /// an `AddressExpired` event yet.
-#[derive(Clone, Debug, PartialEq)]
-pub enum ListenerEvent<TUpgr, TErr> {
-    /// The transport is listening on a new additional [`Multiaddr`].
-    NewAddress(Multiaddr),
-    /// An upgrade, consisting of the upgrade future, the listener address and the remote address.
-    Upgrade {
-        /// The upgrade.
+pub enum TransportEvent<TUpgr, TErr> {
+    /// A new address is being listened on.
+    NewAddress {
+        /// The listener that is listening on the new address.
+        listener_id: ListenerId,
+        /// The new address that is being listened on.
+        listen_addr: Multiaddr,
+    },
+    /// An address is no longer being listened on.
+    AddressExpired {
+        /// The listener that is no longer listening on the address.
+        listener_id: ListenerId,
+        /// The new address that is being listened on.
+        listen_addr: Multiaddr,
+    },
+    /// A connection is incoming on one of the listeners.
+    Incoming {
+        /// The listener that produced the upgrade.
+        listener_id: ListenerId,
+        /// The produced upgrade.
         upgrade: TUpgr,
-        /// The local address which produced this upgrade.
+        /// Local connection address.
         local_addr: Multiaddr,
-        /// The remote address which produced this upgrade.
-        remote_addr: Multiaddr,
+        /// Address used to send back data to the incoming client.
+        send_back_addr: Multiaddr,
+    },
+    /// A listener closed.
+    ListenerClosed {
+        /// The ID of the listener that closed.
+        listener_id: ListenerId,
+        /// Reason for the closure. Contains `Ok(())` if the stream produced `None`, or `Err`
+        /// if the stream produced an error.
+        reason: Result<(), TErr>,
     },
-    /// A [`Multiaddr`] is no longer used for listening.
-    AddressExpired(Multiaddr),
-    /// A non-fatal error has happened on the listener.
+    /// A listener errored.
     ///
-    /// This event should be generated in order to notify the user that something wrong has
-    /// happened. The listener, however, continues to run.
-    Error(TErr),
+    /// The listener will continue to be polled for new events and the event
+    /// is for informational purposes only.
+    Error {
```

Comment:
This is strictly scoped to listener errors, right? How about:

```suggestion
    ListenerError {
```